httpcore5/src/main/java/org/apache/hc/core5/net/URIBuilder.java

```diff
@@ -391,14 +438,16 @@ private String buildString() {
                 formatQuery(sb, this.queryParams, this.charset, false);
             } else if (this.query != null) {
                 sb.append("?");
-                PercentCodec.encode(sb, this.query, this.charset, PercentCodec.URIC, false);
+                PercentCodec.encode(sb, this.query, this.charset,
+                        encodingPolicy == EncodingPolicy.ALL_RESERVED ? PercentCodec.UNRESERVED : PercentCodec.FRAGMENT, false);
```

Comment:
@arturobernalg Careful there. Here we are encoding a complete query that may use sub-delims. Even in `ALL_RESERVED` mode it should still be `URIC`.

I suppose this line should be 
```
encodingPolicy == EncodingPolicy.ALL_RESERVED ? PercentCodec.URIC : PercentCodec.QUERY, false)
```